### PR TITLE
Switching between latent and pixel space training.

### PR DESCRIPTION
**Pull request by Ljackson-su**
This code allows for switching between latent and pixel space training.

### CODE REVIEW

File: mindglass_code/autoencoder.py [hunk 1]
```diff
@@ -81,21 +81,24 @@ def encode_video_latents(self,vid_array):
...
     def decode_video_latents(self, latents):
+        if latents.dim() != 5:
+            latents = latents[None,:]
         latents = 1 / 0.18215 * latents
+        latents = latents.permute(0,2,1,3,4)
         videos = []
-        for i in range(latents.shape[0]):
+        for i in tqdm(range(latents.shape[0]), desc='Decoding samples', total=int(latents.shape[0])):
             with torch.no_grad():
                 video_by_batch = self.model.decode(latents[0])
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:99]: Shouldn't this be `video_by_batch = self.model.decode(latents[i])` instead?

Comment (oxjohanndiep) [reply mindglass_code/autoencoder.py:99]: Otherwise you are just decoding the first batch for the entire loop,

File: mindglass_code/diffusion.py [hunk 2]
```diff
@@ -105,14 +111,18 @@ def p_mean_variance(self, x, t, clip_denoised: bool, cond = None, cond_scale = 1
...
+            # THIS IS DEPENDENT ON THE LATENT SPACE - the best results are seen when this value is clipped to the actual limits of the latent space
             s = 1.
             if self.use_dynamic_thres:
                 s = torch.quantile(rearrange(x_recon, 'b ... -> b (...)').abs(), self.dynamic_thres_percentile, dim = -1)
                 s.clamp_(min = 1.)
                 s = s.view(-1, *((1,) * (x_recon.ndim - 1)))
 
             # Clip by threshold depending on whether static or dynamic.
-            x_recon = x_recon.clamp(-s, s) / s
+            x_recon = x_recon.clamp(-s, s)
```
Comment (oxjohanndiep) [inline mindglass_code/diffusion.py:123]: I guess this is somehow a hack, leaving out the division by s would not follow the exact approach of dynamic thresholding.

File: mindglass_code/main.py [hunk 3]
```diff
@@ -43,15 +43,19 @@
...
+#   - use_autoencoder: True/False. This will modify the size of the input and output space accordingly.
+
 diffusion_model = GaussianDiffusion(
     denoise_fn,
-    image_size = 32,
-    num_frames = 5,
-    channels = 3,
+    image_size = 256,
+    num_frames = 10,
+    channels = 4,
```
Comment (oxjohanndiep) [inline mindglass_code/main.py:52]: It appears to me that `channels` can be set to 3 as default. If we are using autoencoder, `channels `will be replaced by `latent_channels` anyway. If autoencoder is off, having `channels ` set to 4 will lead to errors if the user forgets to set it back to 3.

Comment (Ljackson-su) [reply mindglass_code/main.py:52]: I do not think this will work with the unet as there is no condition inside to set the channels base on if we are using the autoencoder

File: mindglass_code/autoencoder.py (hunk 1 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:99]: Indeed it should!

File: mindglass_code/diffusion.py (hunk 2 above)
Comment (Ljackson-su) [reply mindglass_code/diffusion.py:123]: Exactly. But the division by s lead to the latent space being clamped wrong. Moving forward I assume the latent space will be used more

File: mindglass_code/trainer.py [hunk 4]
```diff
@@ -252,12 +269,18 @@ def train(
...
-
-                all_videos_list = list(map(lambda n: self.ema_model.sample(cond = cond, batch_size=n), batches))
+                
+                # Inference text conditioning must be hard coded.
+                all_videos_list = list(map(lambda n: self.ema_model.sample(cond = ['man plays the piano','man sat at piano'], batch_size=n), batches))
                 all_videos_list = torch.cat(all_videos_list, dim = 0)
 
+                if self.use_autoencoder:
+                    all_videos_list = self.vae.decode_video_latents(all_videos_list)
+                    # Only works when a single video is being used.
```
Comment (oxjohanndiep) [inline mindglass_code/trainer.py:279]: What do you mean by "only works when a single video is being used"?

Comment (oxjohanndiep) [reply mindglass_code/trainer.py:279]: Furthermore, I assume this training only is possible in the overfitting scenario, where we are trying to match the video input. In case the diffusion process produces new creations, the loss for the decoder training does not make any sense?

Comment (oxjohanndiep) [reply mindglass_code/trainer.py:279]: Ah yes, just wanted to be sure :)

File: mindglass_code/main.py (hunk 3 above)
Comment (oxjohanndiep) [reply mindglass_code/main.py:52]: I believe it does, you I already implemented the switch condition between latent and pixel space. 
So the `channels ` affects the `GaussianDiffusion` class. If you decide to set `use_autoencoder` to `True`, `channels` has no meaning at all, since it will be replaced by `latent_channels` everywhere it was used. If however `use_autoencoder` to `False`, `channels` set to 4 would break the code.

Comment (oxjohanndiep) [reply mindglass_code/main.py:52]: But then again, this is just nitpicking, I will merge this so you dont have to do the change.

Comment (Ljackson-su) [reply mindglass_code/main.py:52]: More than happy to change it. Makes sense to keep it clean?

Comment (Ljackson-su) [reply mindglass_code/main.py:52]: I have just had another look and channels in the unet does need to be right, there is no switch in the unet to assign it as 4 if we are using the autoencoder
